### ISSUE TITLE
Meta Transactions PoC

### ISSUE THREAD

**Issue by nick** [OriginProtocol/origin#2001]
Acquiring Eth in order to transact on Origin remains a significant hurdle for sellers. In order for a seller to complete a transaction and get paid by a buyer they must: `createListing`, `acceptOffer` as well as potentially `finalizeOffer` (if the buyer did not finalize), `disputeOffer` and even `transferFrom` if they wish to convert ERC20 to fiat (eg if the transaction was completed in Dai). All of those transactions require gas.

One option is to simply give some free Eth to the seller in order to facilitate these transactions - however this could potentially be abused. Another solution is to use Meta Transactions to pay gas on the user's behalf for approved transactions, such as those on our marketplace contract. We are going to try a PoC for the latter. The basic flow would be as follows:

- User generates a new Ethereum wallet (eg in the mobile app)
- User signs a message proving they own the wallet
- User signs a transaction for creating a listing on Origin marketplace
- User calls centralized Origin endpoint with their wallet address, signed owner proof and signed listing transaction
- Origin performs ecrecover to verify user controls the given wallet
- Origin deploys [ERC725 Proxy](https://github.com/ERC725Alliance/erc725/blob/master/contracts/contracts/Identity.sol) with Meta Transaction support
- Origin changes ERC725 owner to user's wallet
- Origin calls `forward` method on ERC725 with user's signed createListing transaction (these 3 transactions could potentially be combined into 1)
- New listing is created on Marketplace with the user's ERC725 contract as the owner

User now has:

- ERC725 Proxy that they control
- A listing on Origin with their ERC725 Proxy as the owner
- Did not have to fund their wallet

Notes:

- All future seller interactions with Marketplace contract must happen via this ERC725 Proxy
- Gas can either be paid by the user by calling the ERC725 execute method directly, or by 3rd party calling the `forward` method with the user's signed transaction
- Origin does not need to blindly pay for gas on all transactions submitted. Instead we could require certain conditions are met, such as basic KYC, only pay for the user's first 3 listings, etc

For the PoC, I think we just need to:

- Add a `forward` method to the existing [ERC725 v2 implementation ](https://github.com/ERC725Alliance/erc725/blob/master/contracts/contracts/Identity.sol) (example implementation https://github.com/austintgriffith/bouncer-proxy/blob/master/BouncerProxy/BouncerProxy.sol#L77)
- `forward` method should verify signer is ERC725 owner
- Show example of ERC725 being deployed, owner transferred and createListing being executed via `forward` method by 3rd party

Reading material:
- [Meta Transactions](https://medium.com/@austin_48503/ethereum-meta-transactions-90ccf0859e84) explanation
- [Bouncer Proxy](https://github.com/austintgriffith/bouncer-proxy/blob/master/BouncerProxy/BouncerProxy.sol) (example implementation)

**Comment by franckc** [comment]
Thanks for providing all these details @nick - super useful for the rest of the team to see that.

Here are my notes from our chat this morning:
 - Gas cost for deploying proxy contract + making a call to marketplace to create listing ?
   - TBD. We'll find out with Shahul's prototype
 - Aside from proxying the marketplace create listing call, what other calls will we need to proxy ?
   - Update identity
   - Accept offer
 - Real-time fraud analysis for relayer to decide whether or not to accept an incoming request
   - Not needed for prototype, only if we decide to productionize
   - Would live in the relayer service
   - We should be able to check the content of the request. Including data such as the IPFS hash of the listing
   - Before the user can submit a meta transaction to the relayer, we should ask for KYC. At a minimum ask for some attestations (email, etc...)
 - Identity
   - If we deployed a proxy contract for the user it should become their primary address for identity
   - All users won't have a proxy so we'll need to support 2 classes of users: ones with identity address being same as their wallet address, others with identity being their proxy address.
 - What if we decide later to stop supporting meta-transactions ?
   - Simplest would likely to keep support for proxy contract
   - One idea could be to build proxy contract support in a Web3 provider that would abstract if a direct call to the blockchain is made or if the transaction is wrapped and sent to a relayer.
 - Should we support option for the seller to pay for gas using OGN ?
   - Not needed for prototype. Doable. We can decide later.

**Next steps**:
- Shahul started working on prototype
 - We'll organize a meeting for him to demo to the team his prototype
 - This will give us useful data points to decide whether or not we want to move forward and productionize the prototype. In particular:
    - Gas cost the relayer (e.g. Origin) will have to incur for each type of transaction (CreateListing, AcceptListing, UpdateIdentity
    - Scope of the engineering work involved to launch a MVP of meta-transactions

**Comment by shahthepro** [comment]
@franckc 

> Gas cost for deploying proxy contract + making a call to marketplace to create listing ? 

Have pushed a commit to [shahthepro/meta-transactions-poc](https://github.com/OriginProtocol/origin/tree/shahthepro/meta-transactions-poc) that tracks the gas used in the unit tests.

> Aside from proxying the marketplace create listing call, what other calls will we need to proxy ?

We also need to proxy `updateListing()` and `withdrawListing()` methods. Also once a listing has been finalized, we have to provide methods to transfer the funds from user's proxy contract to their own wallet. Or if we modify Marketplace contract, we can directly transfer funds to proxy contract's owner address (if the listing was created with proxy contract).

**Comment by nick** [comment]
Gas price is currently about 857k to deploy an Identity Proxy when the optimize flag is on ($1.20). I think we can get this down significantly by using a [delegate proxy](https://blog.gnosis.pm/solidity-delegateproxy-contracts-e09957d0f201) or similar technique.

**Comment by wanderingstan** [comment]
Interesting additional idea from chatting with Nick this morning; a way for Origin to get paid back:

At same user (seller) signs a transaction for creating a listing on Origin marketplace we could also have also have them sign a transaction transferring money from that wallet to an Origin wallet. This could be an estimate of the gas costs, or even include a small "service charge" for us extending them the loan.

Then, when the user sells their first item and they actually have funds, we can execute those transactions to be repaid.

Basically its like they write us a check for the gas, but we don't cash it until they have money in the bank.

**Comment by shahthepro** [comment]
I have pushed changes to [shahthepro/meta-transactions-poc](https://github.com/OriginProtocol/origin/tree/shahthepro/meta-transactions-poc) branch that allows to use the DApp to create a listing with no ether.

A few points:

1. At present, only `createListing()` method has been added to the relayer. Will have to add other methods.
2. The created proxy address and nonce value of sender are not persisted right now. Will add them to Database. So, once an proxy address has been created for an address, we will have to reuse it. It should be a one-to-one relationship.
3. The `forward()` method doesn't emit events of the called functions (in this case `createListing()`). How are we going to find which call created which listing?
4. Since the entire deploying proxy and forwarding tx process may take long, How about using websockets or doing long-polling?
5. Right now, the flow is like:
      - Put the listing info on IPFS hash
      - Sign the txData
      - Send the signed message and txData to relayer, which submits it to blockchain

    What happens to the IPFS data if the user cancels creating a listing after the first step?
